### ISSUE TITLE
Screen Reader Issues With Modal

### ISSUE THREAD

**Issue by frastlin** [carbon-design-system/carbon-components-react#1659]
## Detailed description

Looking at the [latest version of modal](https://deploy-preview-1657--carbon-components-react.netlify.com/?selectedKind=Modal&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel) there are several issues:

1. Escape does not exit the modal (I don't know if the exit modal is blocked on the example for some reason though)
2. The button does not close the modal (I don't know if the exit modal is blocked on the example for some reason though)
3. The close button is wrapped in a graphic that makes the screen reader announce: "Button Close the modal graphic Close the modal". The graphic should be hidden so the redundant information is not announced.
4. The "label" heading is level 4 and is above the "Modal heading" heading level 2. The "Label" should be heading level 1 or 2 (it doesn't matter), and the "modal heading" should be either the same or deeper than the "label" level.

It may be a good idea to have a button that triggers the modal as when a screen reader clicks on the "Modal" component, they are instantly placed in the modal. It is a little confusing.

## Steps to reproduce the issue

1. Using NVDA go to the [Modal component](https://deploy-preview-1657--carbon-components-react.netlify.com/?selectedKind=Modal&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel)
2. Notice that focus is placed in the modal on page load
3. Hit ctrl+home and arrow down listening to the headings. The levels are out of order.
4. Hit escape
5. Arrow over the exit button and hear the redundant information.
6. hit enter on the exit button and nothing happens.

**Comment by joshblack** [comment]
Hi @frastlin! 👋 Thanks for taking the time to make this issue 😄 

Just to clarify, most examples in our Storybook are meant for demonstrating how to use them in a development context. They should not be used as a place to do accessibility verification testing.

Am going to include one of our accessibility engineers on this topic though just so they can confirm. After they take a look, we can figure out if it's an accessibility violation or something up with our development environment!

cc @dakahn if you have a chance to verify

**Comment by dakahn** [comment]
Hey @frastlin thanks for being patient. Getting to back to this now. So the first two points are weird caveats with using Storybook. In actual implementation that functionality should be working. The last two points are totally valid and need to be addressed.

**Comment by frastlin** [comment]
The problem is that without implementing all these conventions by default, you're expecting that the user of the widget will know these conventions and they don't.
Escape always, in every case, unless a developer specifies otherwise, should close a modal, just as clicking on an accordion button should close the accordion.
Clicking the close button should always close a modal.
The modal should not pop up by default because otherwise the screen reader is stuck in that modal when they click on the link to go to the storybook page with the modal and it is very unexpected.

**Comment by dakahn** [comment]
@frastlin You're exactly right. The demo needs work and should be conveying the modal's actual use and functionality. 👍
